src/dmtest/gendatablocks.py

```diff
@@ -0,0 +1,494 @@
+import copy
+import errno
+import os
+import struct
+
+from pathlib import Path
+from typing import List
+
+# According to pylint numpy should be last
+import numpy
+
+# Each data stream is identified by an 8 character tag
+MAX_TAG_SIZE = 8
+
+# We do math mod 1000 to process the dedupe fraction
+DEDUPE_MODULUS = 1000
+
+def shrink_for_dedupe(number: int, dedupe: float) -> int:
+    """Calculate the block number for the Header.
+
+    In general, if we end up with the same tag, stream number, and block number we will
+    have the same Header and thus the same 'random' data. This is how we generate
+    deduplicated data.
+
+    To accomplish this, the algorithm below maintains the appropriate fraction of unique
+    block numbers. The rest are halved, perhaps multiple times, to map unevenly onto
+    lower (previously used) block numbers
+
+    Parameters
+    ----------
+    number : int
+        The original block number
+    dedupe : float
+        The dedupe rate, scaled up by DEDUPE_MODULUS
+
+    Returns
+    -------
+    int
+        The calculated block number
+
+    """
+    while number > 0 and (number * dedupe) % 1 < dedupe:
+        number = number >> 1
+    return number
+
+class CompareError(Exception):
+    """ Exception raised for all data compare errors """
+    def __init__(self,
+                 message: str,
+                 block_number: int,
+                 actual: bytes,
+                 expected: bytes = None,
+                 byte_number: int = 0):
+        self.block_number = block_number
+        self.actual = actual
+        self.expected = expected
+        self.byte_number = byte_number
+        super().__init__(message)
+
+        def __str__(self) -> str:
+            """ __str__ is to print() the value """
```

Comment:
In the function doc string you don't need to show the name of the function. In fact this function is pretty standard and so the doc string doesn't really add much. Maybe just delete it.